File: lib/larb/mat3.rb
# frozen_string_literal: true

module Larb
  class Mat3
    attr_reader :data

    def initialize(data = nil)
      @data = data || [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0]
    end

    def self.identity
      new
    end

    def self.zero
      new(Array.new(9, 0.0))
    end

    def self.from_mat4(m)
      new([
        m[0], m[1], m[2],
        m[4], m[5], m[6],
        m[8], m[9], m[10]
      ])
    end

    def self.from_mat2d(m)
      new([
        m[0], m[1], 0,
        m[2], m[3], 0,
        m[4], m[5], 1
      ])
    end

    def self.from_quaternion(q)
      x, y, z, w = q.x, q.y, q.z, q.w
      x2 = x + x
      y2 = y + y
      z2 = z + z
      xx = x * x2
      yx = y * x2
      yy = y * y2
      zx = z * x2
      zy = z * y2
      zz = z * z2
      wx = w * x2
      wy = w * y2
      wz = w * z2

      new([
        1 - yy - zz, yx + wz, zx - wy,
        yx - wz, 1 - xx - zz, zy + wx,
        zx + wy, zy - wx, 1 - xx - yy
      ])
    end

    def self.normal_from_mat4(m)
      from_mat4(m).inverse.transpose
    end

    def self.projection(width, height)
      new([
        2.0 / width, 0, 0,
        0, -2.0 / height, 0,
        -1, 1, 1
      ])
    end

    def self.translation(x, y)
      new([
        1, 0, 0,
        0, 1, 0,
        x.to_f, y.to_f, 1
      ])
    end

    def self.rotation(radians)
      c = Math.cos(radians)
      s = Math.sin(radians)
      new([
        c, s, 0,
        -s, c, 0,
        0, 0, 1
      ])
    end

    def self.scaling(x, y)
      new([
        x.to_f, 0, 0,
        0, y.to_f, 0,
        0, 0, 1
      ])
    end

    def [](i)
      @data[i]
    end

    def []=(i, v)
      @data[i] = v.to_f
    end

    def *(other)
      case other
      when Mat3
        a = @data
        b = other.data
        Mat3.new([
          a[0] * b[0] + a[3] * b[1] + a[6] * b[2],
          a[1] * b[0] + a[4] * b[1] + a[7] * b[2],
          a[2] * b[0] + a[5] * b[1] + a[8] * b[2],
          a[0] * b[3] + a[3] * b[4] + a[6] * b[5],
          a[1] * b[3] + a[4] * b[4] + a[7] * b[5],
          a[2] * b[3] + a[5] * b[4] + a[8] * b[5],
          a[0] * b[6] + a[3] * b[7] + a[6] * b[8],
          a[1] * b[6] + a[4] * b[7] + a[7] * b[8],
          a[2] * b[6] + a[5] * b[7] + a[8] * b[8]
        ])
      when Vec3
        Vec3.new(
          @data[0] * other.x + @data[3] * other.y + @data[6] * other.z,
          @data[1] * other.x + @data[4] * other.y + @data[7] * other.z,
          @data[2] * other.x + @data[5] * other.y + @data[8] * other.z
        )
      when Numeric
        Mat3.new(@data.map { |v| v * other })
      end
    end

    def +(other)
      Mat3.new(@data.zip(other.data).map { |a, b| a + b })
    end

    def -(other)
      Mat3.new(@data.zip(other.data).map { |a, b| a - b })
    end

    def determinant
      a = @data
      a[0] * (a[4] * a[8] - a[5] * a[7]) -
        a[3] * (a[1] * a[8] - a[2] * a[7]) +
        a[6] * (a[1] * a[5] - a[2] * a[4])
    end

    def inverse
      a = @data
      det = determinant
      raise "Matrix is not invertible" if det.abs < 1e-10

      inv_det = 1.0 / det
      Mat3.new([
        (a[4] * a[8] - a[5] * a[7]) * inv_det,
        (a[2] * a[7] - a[1] * a[8]) * inv_det,
        (a[1] * a[5] - a[2] * a[4]) * inv_det,
        (a[5] * a[6] - a[3] * a[8]) * inv_det,
        (a[0] * a[8] - a[2] * a[6]) * inv_det,
        (a[2] * a[3] - a[0] * a[5]) * inv_det,
        (a[3] * a[7] - a[4] * a[6]) * inv_det,
        (a[1] * a[6] - a[0] * a[7]) * inv_det,
        (a[0] * a[4] - a[1] * a[3]) * inv_det
      ])
    end

    def transpose
      Mat3.new([
        @data[0], @data[3], @data[6],
        @data[1], @data[4], @data[7],
        @data[2], @data[5], @data[8]
      ])
    end

    def adjoint
      a = @data
      Mat3.new([
        a[4] * a[8] - a[5] * a[7],
        a[2] * a[7] - a[1] * a[8],
        a[1] * a[5] - a[2] * a[4],
        a[5] * a[6] - a[3] * a[8],
        a[0] * a[8] - a[2] * a[6],
        a[2] * a[3] - a[0] * a[5],
        a[3] * a[7] - a[4] * a[6],
        a[1] * a[6] - a[0] * a[7],
        a[0] * a[4] - a[1] * a[3]
      ])
    end

    def frobenius_norm
      Math.sqrt(@data.sum { |v| v * v })
    end

    def translate(x, y)
      a = @data
      Mat3.new([
        a[0], a[1], a[2],
        a[3], a[4], a[5],
        x * a[0] + y * a[3] + a[6],
        x * a[1] + y * a[4] + a[7],
        x * a[2] + y * a[5] + a[8]
      ])
    end

    def rotate(radians)
      self * Mat3.rotation(radians)
    end

    def scale(x, y)
      a = @data
      Mat3.new([
        a[0] * x, a[1] * x, a[2] * x,
        a[3] * y, a[4] * y, a[5] * y,
        a[6], a[7], a[8]
      ])
    end

    def to_a
      @data.dup
    end

    def ==(other)
      return false unless other.is_a?(Mat3)

      @data == other.data
    end

    def near?(other, epsilon = 1e-6)
      @data.zip(other.data).all? { |a, b| (a - b).abs < epsilon }
    end

    def inspect
      "Mat3[\n" \
      "  #{@data[0..2].map { |v| format('%8.4f', v) }.join(', ')}\n" \
      "  #{@data[3..5].map { |v| format('%8.4f', v) }.join(', ')}\n" \
      "  #{@data[6..8].map { |v| format('%8.4f', v) }.join(', ')}\n]"
    end

    alias to_s inspect
  end
end

File: ext/larb/mat3.c
#include "mat3.h"

#include <math.h>

static void mat3_free(void *ptr) {
  xfree(ptr);
}

static size_t mat3_memsize(const void *ptr) {
  return sizeof(Mat3Data);
}

static const rb_data_type_t mat3_type = {
    "Mat3",
    {0, mat3_free, mat3_memsize},
    0,
    0,
    RUBY_TYPED_FREE_IMMEDIATELY,
};

static VALUE cMat3 = Qnil;
static VALUE cVec3 = Qnil;

static double value_to_double(VALUE value) {
  VALUE coerced = rb_funcall(value, rb_intern("to_f"), 0);
  return NUM2DBL(coerced);
}

static Mat3Data *mat3_get(VALUE obj) {
  Mat3Data *data = NULL;
  TypedData_Get_Struct(obj, Mat3Data, &mat3_type, data);
  return data;
}

static VALUE mat3_build(VALUE klass, const double *values) {
  VALUE obj = mat3_alloc(klass);
  Mat3Data *data = mat3_get(obj);
  for (int i = 0; i < 9; i++) {
    data->data[i] = values[i];
  }
  return obj;
}

static VALUE mat3_build9(VALUE klass, double v0, double v1, double v2,
                         double v3, double v4, double v5, double v6,
                         double v7, double v8) {
  double values[9] = {v0, v1, v2, v3, v4, v5, v6, v7, v8};
  return mat3_build(klass, values);
}

VALUE mat3_alloc(VALUE klass) {
  Mat3Data *data = ALLOC(Mat3Data);
  data->data[0] = 1.0;
  data->data[1] = 0.0;
  data->data[2] = 0.0;
  data->data[3] = 0.0;
  data->data[4] = 1.0;
  data->data[5] = 0.0;
  data->data[6] = 0.0;
  data->data[7] = 0.0;
  data->data[8] = 1.0;
  return TypedData_Wrap_Struct(klass, &mat3_type, data);
}

VALUE mat3_initialize(int argc, VALUE *argv, VALUE self) {
  VALUE data_arg = Qnil;
  Mat3Data *data = mat3_get(self);

  rb_scan_args(argc, argv, "01", &data_arg);
  if (NIL_P(data_arg)) {
    data->data[0] = 1.0;
    data->data[1] = 0.0;
    data->data[2] = 0.0;
    data->data[3] = 0.0;
    data->data[4] = 1.0;
    data->data[5] = 0.0;
    data->data[6] = 0.0;
    data->data[7] = 0.0;
    data->data[8] = 1.0;
    return self;
  }

  VALUE ary = rb_check_array_type(data_arg);
  if (NIL_P(ary)) {
    rb_raise(rb_eTypeError, "expected Array");
  }

  for (int i = 0; i < 9; i++) {
    data->data[i] = value_to_double(rb_ary_entry(ary, i));
  }

  return self;
}

static VALUE mat3_class_identity(VALUE klass) {
  return mat3_alloc(klass);
}

static VALUE mat3_class_zero(VALUE klass) {
  return mat3_build9(klass, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0);
}

static VALUE mat3_class_from_mat4(VALUE klass, VALUE mat4) {
  double values[9];
  values[0] = value_to_double(rb_funcall(mat4, rb_intern("[]"), 1, INT2NUM(0)));
  values[1] = value_to_double(rb_funcall(mat4, rb_intern("[]"), 1, INT2NUM(1)));
  values[2] = value_to_double(rb_funcall(mat4, rb_intern("[]"), 1, INT2NUM(2)));
  values[3] = value_to_double(rb_funcall(mat4, rb_intern("[]"), 1, INT2NUM(4)));
  values[4] = value_to_double(rb_funcall(mat4, rb_intern("[]"), 1, INT2NUM(5)));
  values[5] = value_to_double(rb_funcall(mat4, rb_intern("[]"), 1, INT2NUM(6)));
  values[6] = value_to_double(rb_funcall(mat4, rb_intern("[]"), 1, INT2NUM(8)));
  values[7] = value_to_double(rb_funcall(mat4, rb_intern("[]"), 1, INT2NUM(9)));
  values[8] = value_to_double(rb_funcall(mat4, rb_intern("[]"), 1, INT2NUM(10)));
  return mat3_build(klass, values);
}

static VALUE mat3_class_from_mat2d(VALUE klass, VALUE mat2d) {
  double values[9];
  values[0] = value_to_double(rb_funcall(mat2d, rb_intern("[]"), 1, INT2NUM(0)));
  values[1] = value_to_double(rb_funcall(mat2d, rb_intern("[]"), 1, INT2NUM(1)));
  values[2] = 0.0;
  values[3] = value_to_double(rb_funcall(mat2d, rb_intern("[]"), 1, INT2NUM(2)));
  values[4] = value_to_double(rb_funcall(mat2d, rb_intern("[]"), 1, INT2NUM(3)));
  values[5] = 0.0;
  values[6] = value_to_double(rb_funcall(mat2d, rb_intern("[]"), 1, INT2NUM(4)));
  values[7] = value_to_double(rb_funcall(mat2d, rb_intern("[]"), 1, INT2NUM(5)));
  values[8] = 1.0;
  return mat3_build(klass, values);
}

static VALUE mat3_class_from_quaternion(VALUE klass, VALUE quat) {
  double x = value_to_double(rb_funcall(quat, rb_intern("x"), 0));
  double y = value_to_double(rb_funcall(quat, rb_intern("y"), 0));
  double z = value_to_double(rb_funcall(quat, rb_intern("z"), 0));
  double w = value_to_double(rb_funcall(quat, rb_intern("w"), 0));
  double x2 = x + x;
  double y2 = y + y;
  double z2 = z + z;
  double xx = x * x2;
  double yx = y * x2;
  double yy = y * y2;
  double zx = z * x2;
  double zy = z * y2;
  double zz = z * z2;
  double wx = w * x2;
  double wy = w * y2;
  double wz = w * z2;

  return mat3_build9(klass, 1 - yy - zz, yx + wz, zx - wy, yx - wz,
                     1 - xx - zz, zy + wx, zx + wy, zy - wx,
                     1 - xx - yy);
}

static VALUE mat3_class_normal_from_mat4(VALUE klass, VALUE mat4) {
  VALUE m3 = mat3_class_from_mat4(klass, mat4);
  VALUE inv = rb_funcall(m3, rb_intern("inverse"), 0);
  return rb_funcall(inv, rb_intern("transpose"), 0);
}

static VALUE mat3_class_projection(VALUE klass, VALUE width, VALUE height) {
  double w = value_to_double(width);
  double h = value_to_double(height);
  return mat3_build9(klass, 2.0 / w, 0.0, 0.0, 0.0, -2.0 / h, 0.0, -1.0, 1.0,
                     1.0);
}

static VALUE mat3_class_translation(VALUE klass, VALUE x, VALUE y) {
  return mat3_build9(klass, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0, value_to_double(x),
                     value_to_double(y), 1.0);
}

static VALUE mat3_class_rotation(VALUE klass, VALUE radians) {
  double r = value_to_double(radians);
  double c = cos(r);
  double s = sin(r);
  return mat3_build9(klass, c, s, 0.0, -s, c, 0.0, 0.0, 0.0, 1.0);
}

static VALUE mat3_class_scaling(VALUE klass, VALUE x, VALUE y) {
  return mat3_build9(klass, value_to_double(x), 0.0, 0.0, 0.0,
                     value_to_double(y), 0.0, 0.0, 0.0, 1.0);
}

VALUE mat3_aref(VALUE self, VALUE index) {
  Mat3Data *data = mat3_get(self);
  long idx = NUM2LONG(index);
  if (idx < 0 || idx > 8) {
    return Qnil;
  }
  return DBL2NUM(data->data[idx]);
}

VALUE mat3_aset(VALUE self, VALUE index, VALUE value) {
  Mat3Data *data = mat3_get(self);
  long idx = NUM2LONG(index);
  if (idx < 0 || idx > 8) {
    rb_raise(rb_eIndexError, "index %ld out of range", idx);
  }
  data->data[idx] = value_to_double(value);
  return value;
}

VALUE mat3_mul(VALUE self, VALUE other) {
  Mat3Data *a = mat3_get(self);

  if (rb_obj_is_kind_of(other, cMat3)) {
    Mat3Data *b = mat3_get(other);
    return mat3_build9(
        rb_obj_class(self),
        a->data[0] * b->data[0] + a->data[3] * b->data[1] +
            a->data[6] * b->data[2],
        a->data[1] * b->data[0] + a->data[4] * b->data[1] +
            a->data[7] * b->data[2],
        a->data[2] * b->data[0] + a->data[5] * b->data[1] +
            a->data[8] * b->data[2],
        a->data[0] * b->data[3] + a->data[3] * b->data[4] +
            a->data[6] * b->data[5],
        a->data[1] * b->data[3] + a->data[4] * b->data[4] +
            a->data[7] * b->data[5],
        a->data[2] * b->data[3] + a->data[5] * b->data[4] +
            a->data[8] * b->data[5],
        a->data[0] * b->data[6] + a->data[3] * b->data[7] +
            a->data[6] * b->data[8],
        a->data[1] * b->data[6] + a->data[4] * b->data[7] +
            a->data[7] * b->data[8],
        a->data[2] * b->data[6] + a->data[5] * b->data[7] +
            a->data[8] * b->data[8]);
  }

  if (rb_obj_is_kind_of(other, cVec3)) {
    double x = value_to_double(rb_funcall(other, rb_intern("x"), 0));
    double y = value_to_double(rb_funcall(other, rb_intern("y"), 0));
    double z = value_to_double(rb_funcall(other, rb_intern("z"), 0));
    VALUE vec3_class = rb_const_get(mLarb, rb_intern("Vec3"));
    return rb_funcall(vec3_class, rb_intern("new"), 3,
                      DBL2NUM(a->data[0] * x + a->data[3] * y +
                              a->data[6] * z),
                      DBL2NUM(a->data[1] * x + a->data[4] * y +
                              a->data[7] * z),
                      DBL2NUM(a->data[2] * x + a->data[5] * y +
                              a->data[8] * z));
  }

  if (rb_obj_is_kind_of(other, rb_cNumeric)) {
    double s = value_to_double(other);
    return mat3_build9(rb_obj_class(self), a->data[0] * s, a->data[1] * s,
                       a->data[2] * s, a->data[3] * s, a->data[4] * s,
                       a->data[5] * s, a->data[6] * s, a->data[7] * s,
                       a->data[8] * s);
  }

  return Qnil;
}

VALUE mat3_add(VALUE self, VALUE other) {
  Mat3Data *a = mat3_get(self);
  Mat3Data *b = mat3_get(other);
  return mat3_build9(rb_obj_class(self), a->data[0] + b->data[0],
                     a->data[1] + b->data[1], a->data[2] + b->data[2],
                     a->data[3] + b->data[3], a->data[4] + b->data[4],
                     a->data[5] + b->data[5], a->data[6] + b->data[6],
                     a->data[7] + b->data[7], a->data[8] + b->data[8]);
}

VALUE mat3_sub(VALUE self, VALUE other) {
  Mat3Data *a = mat3_get(self);
  Mat3Data *b = mat3_get(other);
  return mat3_build9(rb_obj_class(self), a->data[0] - b->data[0],
                     a->data[1] - b->data[1], a->data[2] - b->data[2],
                     a->data[3] - b->data[3], a->data[4] - b->data[4],
                     a->data[5] - b->data[5], a->data[6] - b->data[6],
                     a->data[7] - b->data[7], a->data[8] - b->data[8]);
}

VALUE mat3_determinant(VALUE self) {
  Mat3Data *a = mat3_get(self);
  double det = a->data[0] * (a->data[4] * a->data[8] - a->data[5] * a->data[7]) -
               a->data[3] * (a->data[1] * a->data[8] - a->data[2] * a->data[7]) +
               a->data[6] * (a->data[1] * a->data[5] - a->data[2] * a->data[4]);
  return DBL2NUM(det);
}

VALUE mat3_inverse(VALUE self) {
  Mat3Data *a = mat3_get(self);
  double det = a->data[0] * (a->data[4] * a->data[8] - a->data[5] * a->data[7]) -
               a->data[3] * (a->data[1] * a->data[8] - a->data[2] * a->data[7]) +
               a->data[6] * (a->data[1] * a->data[5] - a->data[2] * a->data[4]);
  if (fabs(det) < 1e-10) {
    rb_raise(rb_eRuntimeError, "Matrix is not invertible");
  }
  double inv_det = 1.0 / det;
  return mat3_build9(
      rb_obj_class(self),
      (a->data[4] * a->data[8] - a->data[5] * a->data[7]) * inv_det,
      (a->data[2] * a->data[7] - a->data[1] * a->data[8]) * inv_det,
      (a->data[1] * a->data[5] - a->data[2] * a->data[4]) * inv_det,
      (a->data[5] * a->data[6] - a->data[3] * a->data[8]) * inv_det,
      (a->data[0] * a->data[8] - a->data[2] * a->data[6]) * inv_det,
      (a->data[2] * a->data[3] - a->data[0] * a->data[5]) * inv_det,
      (a->data[3] * a->data[7] - a->data[4] * a->data[6]) * inv_det,
      (a->data[1] * a->data[6] - a->data[0] * a->data[7]) * inv_det,
      (a->data[0] * a->data[4] - a->data[1] * a->data[3]) * inv_det);
}

VALUE mat3_transpose(VALUE self) {
  Mat3Data *a = mat3_get(self);
  return mat3_build9(rb_obj_class(self), a->data[0], a->data[3], a->data[6],
                     a->data[1], a->data[4], a->data[7], a->data[2],
                     a->data[5], a->data[8]);
}

VALUE mat3_adjoint(VALUE self) {
  Mat3Data *a = mat3_get(self);
  return mat3_build9(
      rb_obj_class(self), a->data[4] * a->data[8] - a->data[5] * a->data[7],
      a->data[2] * a->data[7] - a->data[1] * a->data[8],
      a->data[1] * a->data[5] - a->data[2] * a->data[4],
      a->data[5] * a->data[6] - a->data[3] * a->data[8],
      a->data[0] * a->data[8] - a->data[2] * a->data[6],
      a->data[2] * a->data[3] - a->data[0] * a->data[5],
      a->data[3] * a->data[7] - a->data[4] * a->data[6],
      a->data[1] * a->data[6] - a->data[0] * a->data[7],
      a->data[0] * a->data[4] - a->data[1] * a->data[3]);
}

VALUE mat3_frobenius_norm(VALUE self) {
  Mat3Data *a = mat3_get(self);
  double sum = 0.0;
  for (int i = 0; i < 9; i++) {
    sum += a->data[i] * a->data[i];
  }
  return DBL2NUM(sqrt(sum));
}

VALUE mat3_translate(VALUE self, VALUE x, VALUE y) {
  Mat3Data *a = mat3_get(self);
  double dx = value_to_double(x);
  double dy = value_to_double(y);
  return mat3_build9(rb_obj_class(self), a->data[0], a->data[1], a->data[2],
                     a->data[3], a->data[4], a->data[5],
                     dx * a->data[0] + dy * a->data[3] + a->data[6],
                     dx * a->data[1] + dy * a->data[4] + a->data[7],
                     dx * a->data[2] + dy * a->data[5] + a->data[8]);
}

VALUE mat3_rotate(VALUE self, VALUE radians) {
  VALUE rot = mat3_class_rotation(cMat3, radians);
  return mat3_mul(self, rot);
}

VALUE mat3_scale(VALUE self, VALUE x, VALUE y) {
  Mat3Data *a = mat3_get(self);
  double sx = value_to_double(x);
  double sy = value_to_double(y);
  return mat3_build9(rb_obj_class(self), a->data[0] * sx, a->data[1] * sx,
                     a->data[2] * sx, a->data[3] * sy, a->data[4] * sy,
                     a->data[5] * sy, a->data[6], a->data[7], a->data[8]);
}

VALUE mat3_to_a(VALUE self) {
  Mat3Data *a = mat3_get(self);
  VALUE ary = rb_ary_new_capa(9);
  for (int i = 0; i < 9; i++) {
    rb_ary_push(ary, DBL2NUM(a->data[i]));
  }
  return ary;
}

VALUE mat3_data(VALUE self) {
  return mat3_to_a(self);
}

VALUE mat3_equal(VALUE self, VALUE other) {
  if (!rb_obj_is_kind_of(other, cMat3)) {
    return Qfalse;
  }
  Mat3Data *a = mat3_get(self);
  Mat3Data *b = mat3_get(other);
  for (int i = 0; i < 9; i++) {
    if (a->data[i] != b->data[i]) {
      return Qfalse;
    }
  }
  return Qtrue;
}

VALUE mat3_near(int argc, VALUE *argv, VALUE self) {
  VALUE other = Qnil;
  VALUE epsilon = Qnil;

  rb_scan_args(argc, argv, "11", &other, &epsilon);
  Mat3Data *a = mat3_get(self);
  Mat3Data *b = mat3_get(other);
  double eps = NIL_P(epsilon) ? 1e-6 : value_to_double(epsilon);

  for (int i = 0; i < 9; i++) {
    if (fabs(a->data[i] - b->data[i]) >= eps) {
      return Qfalse;
    }
  }
  return Qtrue;
}

static VALUE mat3_format_value(double value) {
  return rb_funcall(rb_mKernel, rb_intern("format"), 2,
                    rb_str_new_cstr("%8.4f"), DBL2NUM(value));
}

static VALUE mat3_row_string(const double *values, int offset) {
  VALUE str = mat3_format_value(values[offset]);
  for (int i = 1; i < 3; i++) {
    rb_str_cat_cstr(str, ", ");
    rb_str_concat(str, mat3_format_value(values[offset + i]));
  }
  return str;
}

VALUE mat3_inspect(VALUE self) {
  Mat3Data *a = mat3_get(self);
  VALUE str = rb_str_new_cstr("Mat3[\n  ");
  rb_str_concat(str, mat3_row_string(a->data, 0));
  rb_str_cat_cstr(str, "\n  ");
  rb_str_concat(str, mat3_row_string(a->data, 3));
  rb_str_cat_cstr(str, "\n  ");
  rb_str_concat(str, mat3_row_string(a->data, 6));
  rb_str_cat_cstr(str, "\n]");
  return str;
}

void Init_mat3(VALUE module) {
  cMat3 = rb_define_class_under(module, "Mat3", rb_cObject);
  cVec3 = rb_const_get(mLarb, rb_intern("Vec3"));

  rb_define_alloc_func(cMat3, mat3_alloc);
  rb_define_method(cMat3, "initialize", mat3_initialize, -1);

  rb_define_singleton_method(cMat3, "identity", mat3_class_identity, 0);
  rb_define_singleton_method(cMat3, "zero", mat3_class_zero, 0);
  rb_define_singleton_method(cMat3, "from_mat4", mat3_class_from_mat4, 1);
  rb_define_singleton_method(cMat3, "from_mat2d", mat3_class_from_mat2d, 1);
  rb_define_singleton_method(cMat3, "from_quaternion",
                             mat3_class_from_quaternion, 1);
  rb_define_singleton_method(cMat3, "normal_from_mat4",
                             mat3_class_normal_from_mat4, 1);
  rb_define_singleton_method(cMat3, "projection", mat3_class_projection, 2);
  rb_define_singleton_method(cMat3, "translation", mat3_class_translation, 2);
  rb_define_singleton_method(cMat3, "rotation", mat3_class_rotation, 1);
  rb_define_singleton_method(cMat3, "scaling", mat3_class_scaling, 2);

  rb_define_method(cMat3, "data", mat3_data, 0);
  rb_define_method(cMat3, "[]", mat3_aref, 1);
  rb_define_method(cMat3, "[]=", mat3_aset, 2);
  rb_define_method(cMat3, "*", mat3_mul, 1);
  rb_define_method(cMat3, "+", mat3_add, 1);
  rb_define_method(cMat3, "-", mat3_sub, 1);
  rb_define_method(cMat3, "determinant", mat3_determinant, 0);
  rb_define_method(cMat3, "inverse", mat3_inverse, 0);
  rb_define_method(cMat3, "transpose", mat3_transpose, 0);
  rb_define_method(cMat3, "adjoint", mat3_adjoint, 0);
  rb_define_method(cMat3, "frobenius_norm", mat3_frobenius_norm, 0);
  rb_define_method(cMat3, "translate", mat3_translate, 2);
  rb_define_method(cMat3, "rotate", mat3_rotate, 1);
  rb_define_method(cMat3, "scale", mat3_scale, 2);
  rb_define_method(cMat3, "to_a", mat3_to_a, 0);
  rb_define_method(cMat3, "==", mat3_equal, 1);
  rb_define_method(cMat3, "near?", mat3_near, -1);
  rb_define_method(cMat3, "inspect", mat3_inspect, 0);
  rb_define_alias(cMat3, "to_s", "inspect");
}
